# HG changeset patch
# User David Meyer <pdox@google.com>
# Date 1301681159 25200
# Branch pnacl-sfi
# Node ID eda43d7b2ccbd1ce596ca51be0fb14d9dc686e56
# Parent 9ac0cbb3d8cfc83303f7b612c15856719a0203c1
Fix off-by-one error in topological sort adjustment in X86-64 SelectAddr

 From llvm-pnacl-0022-246-eda43d7b2ccbd1ce596ca51be0fb14d9dc686e56.patch

 From .hg/patches/llvm-pnacl-0022-246-eda43d7b2ccbd1ce596ca51be0fb14d9dc686e56.patch.stripped

diff -r 9ac0cbb3d8cf lib/Target/X86/X86ISelDAGToDAG.cpp
--- a/lib/Target/X86/X86ISelDAGToDAG.cpp	Thu Mar 31 14:19:20 2011 -0700
+++ b/lib/Target/X86/X86ISelDAGToDAG.cpp	Wed Jun 08 16:38:50 2011 -0700
@@ -1237,10 +1237,11 @@
     if (Index.getValueType() != MVT::i64) {
       Index = CurDAG->getZExtOrTrunc(Index, Index.getDebugLoc(), MVT::i64);
       // Insert the new node into the topological ordering.
-      if (Index->getNodeId() == -1 ||
-          Index->getNodeId() > N.getNode()->getNodeId()) {
-        CurDAG->RepositionNode(N.getNode(), Index.getNode());
-        Index->setNodeId(N.getNode()->getNodeId());
+      if (Parent &&
+          (Index->getNodeId() == -1 ||
+           Index->getNodeId() > Parent->getNodeId())) {
+        CurDAG->RepositionNode(Parent, Index.getNode());
+        Index->setNodeId(Parent->getNodeId());
       }
     }
   }
